src/clients/MultiCallerClient.ts

```diff
@@ -40,29 +42,167 @@ export const unknownRevertReasonMethodsToIgnore = new Set([
 
 export class MultiCallerClient {
   private transactions: AugmentedTransaction[] = [];
+  private newMulticaller = false;
   protected txnClient: TransactionClient;
+  protected txns: { [chainId: number]: AugmentedTransaction[] } = {};
+  protected valueTxns: { [chainId: number]: AugmentedTransaction[] } = {};
+  // Legacy mode is a temporary feature to support transition to the updated multicaller implementation.
   constructor(
     readonly logger: winston.Logger,
-    readonly chunkSize: { [chainId: number]: number } = DEFAULT_CHAIN_MULTICALL_CHUNK_SIZE
+    readonly chunkSize: { [chainId: number]: number } = DEFAULT_CHAIN_MULTICALL_CHUNK_SIZE,
+    newMulticaller = false
   ) {
+    this.newMulticaller = newMulticaller || process.env.NEW_MULTICALLER === "true";
     this.txnClient = new TransactionClient(logger);
   }
 
-  // Adds all information associated with a transaction to the transaction queue. This is the intention of the
-  // caller to send a transaction. The transaction might not be executable, which should be filtered later.
-  enqueueTransaction(transaction: AugmentedTransaction) {
-    this.transactions.push(transaction);
+  // Adds all information associated with a transaction to the transaction queue.
+  enqueueTransaction(txn: AugmentedTransaction) {
+    if (this.newMulticaller) {
+      // Value transactions are sorted immediately because the UMA multicall implementation rejects them.
+      const txnQueue = txn.value && txn.value.gt(0) ? this.valueTxns : this.txns;
+      if (txnQueue[txn.chainId] === undefined) txnQueue[txn.chainId] = [];
+      txnQueue[txn.chainId].push(txn);
+    } else this.transactions.push(txn);
   }
 
   transactionCount() {
-    return this.transactions.length;
+    let nTxns = this.transactions.length;
+    Object.values(this.txns).forEach((txnQueue) => (nTxns += txnQueue.length));
+    Object.values(this.valueTxns).forEach((txnQueue) => (nTxns += txnQueue.length));
+    return nTxns;
   }
 
-  clearTransactionQueue() {
-    this.transactions = [];
+  clearTransactionQueue(chainId: number = null) {
+    if (this.newMulticaller) {
+      if (chainId) {
```

Comment:
```suggestion
      if (chainId !== null) {
```

to handle chain ID 0

Reply:
Seems unlike that we'd ever have a chainId of 0, but it'd also be weird if a caller passed in a chainId of 0 and we flushed everything. Addressed here: https://github.com/across-protocol/relayer-v2/pull/424/commits/c6d85c4b792e8179925dc27548f2118363e676e7